api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxCurrCohortQueries.java

```diff
@@ -723,4 +728,171 @@ public CohortDefinition getPatientWhoAfterMostRecentDateHaveDrusPickupOrConsulta
+    cd.addParameter(new Parameter("onOrAfter", "After Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    CohortDefinition pickupDiff = getPatientsWithNextPickup83to173Days();
+    CohortDefinition quarterlyDispensation = getPatientsWithQuarterlyTypeOfDispensation();
+    CohortDefinition startOrContinue = getPatientsWithStartOrContinueOnQuarterlyDispensation();
+    CohortDefinition completed = getPatientsWithCompletedOnQuarterlyDispensation();
+    cd.addSearch("pickupDiff", mapStraightThrough(pickupDiff));
+    String codedObsMappings =
+        "onOrAfter=${onOrAfter},onOrBefore=${onOrBefore},locationList=${location}";
+    cd.addSearch("dispensation", EptsReportUtils.map(quarterlyDispensation, codedObsMappings));
+    cd.addSearch("startOrContinue", EptsReportUtils.map(startOrContinue, codedObsMappings));
+    cd.addSearch("completed", EptsReportUtils.map(completed, codedObsMappings));
+    cd.setCompositionString("(pickupDiff OR dispensation OR startOrContinue) NOT completed");
+    return cd;
+  }
+
+  public CohortDefinition getPatientsWithMoreThan6MonthsOfDispensationQuantity() {
+    CompositionCohortDefinition cd = new CompositionCohortDefinition();
+    cd.addParameter(new Parameter("onOrBefore", "Before Date", Date.class));
+    cd.addParameter(new Parameter("onOrAfter", "After Date", Date.class));
+    cd.addParameter(new Parameter("location", "Location", Location.class));
+    CohortDefinition pickupDiff = getPatientsWithNextPickupMoreThan173Days();
+    CohortDefinition quarterlyDispensation = getPatientsWithQuarterlyTypeOfDispensation();
+    CohortDefinition startOrContinue = getPatientsWithStartOrContinueOnQuarterlyDispensation();
+    CohortDefinition completed = getPatientsWithCompletedOnQuarterlyDispensation();
+    cd.addSearch("pickupDiff", mapStraightThrough(pickupDiff));
+    String codedObsMappings =
+        "onOrAfter=${onOrAfter},onOrBefore=${onOrBefore},locationList=${location}";
+    cd.addSearch("dispensation", EptsReportUtils.map(quarterlyDispensation, codedObsMappings));
+    cd.addSearch("startOrContinue", EptsReportUtils.map(startOrContinue, codedObsMappings));
+    cd.addSearch("completed", EptsReportUtils.map(completed, codedObsMappings));
+    cd.setCompositionString("(pickupDiff OR dispensation OR startOrContinue) NOT completed");
+    return cd;
+  }
+
+  private CohortDefinition getPatientsWithNextPickupMoreThan173Days() {
+    return getPatientsWithNextPickupBetweenDaysAfterLastPharmacyEncounter(173, null);
+  }
+
+  private CohortDefinition getPatientsWithNextPickup83to173Days() {
+    return getPatientsWithNextPickupBetweenDaysAfterLastPharmacyEncounter(83, 173);
+  }
+
+  private CohortDefinition getPatientsWithNextPickupLessThan83days() {
+    return getPatientsWithNextPickupBetweenDaysAfterLastPharmacyEncounter(null, 83);
+  }
+
+  private CohortDefinition getPatientsWithMonthlyTypeOfDispensation() {
```

Comment:
Sometimes back we realized the reporting ORM was returning different values. Can you verify the results of this definition returns the same values as the SQL version? I know here the obs_datetime is used when filtering the last record, but when you looking for the last encounter, you will end up applying the encounter datetime, assuming the encounter date and obs date are similar for everyone, this might NOT have effect, but what if those are different? are we applying encounter date or obs date when filtering the last entry? Please ask

Reply:
For `FICHA CLINICA` form which uses `S.TARV: ADULTO SEGUIMENTO` encounter_type `encounter_datetime` is the same as `obs_datetime`.